src/libraries/System.Text.Json/gen/JsonSourceGenerator.Parser.cs

```diff
@@ -0,0 +1,523 @@
+                            hasSetter = PropertyAccessorCanBeReferenced(setMethod, hasJsonInclude) && !setMethod.IsInitOnly();
+                            getterIsVirtual = propertyInfo.GetMethod?.IsVirtual == true;
+                            setterIsVirtual = propertyInfo.SetMethod?.IsVirtual == true;
+                        }
+                        break;
+                    case FieldInfo fieldInfo:
+                        {
+                            Debug.Assert(fieldInfo.IsPublic);
+
+                            memberCLRType = fieldInfo.FieldType;
+                            hasGetter = true;
+                            hasSetter = !fieldInfo.IsInitOnly;
+                        }
+                        break;
+                    default:
+                        throw new InvalidOperationException();
+                }
+
+                return new PropertyMetadata
+                {
+                    ClrName = memberInfo.Name,
+                    IsProperty = memberInfo.MemberType == MemberTypes.Property,
+                    JsonPropertyName = jsonPropertyName,
+                    HasGetter = hasGetter,
+                    HasSetter = hasSetter,
+                    GetterIsVirtual = getterIsVirtual,
+                    SetterIsVirtual = setterIsVirtual,
+                    IgnoreCondition = ignoreCondition,
+                    NumberHandling = numberHandling,
+                    HasJsonInclude = hasJsonInclude,
+                    TypeMetadata = GetOrAddTypeMetadata(memberCLRType),
+                    DeclaringTypeCompilableName = memberInfo.DeclaringType.GetUniqueCompilableTypeName(),
+                    ConverterInstantiationLogic = converterInstantiationLogic
+                };
+            }
+
+            private static bool PropertyAccessorCanBeReferenced(MethodInfo? memberAccessor, bool hasJsonInclude) =>
+                (memberAccessor != null && !memberAccessor.IsPrivate) && (memberAccessor.IsPublic || hasJsonInclude);
+
+            private string? GetConverterInstantiationLogic(CustomAttributeData attributeData)
+            {
+                if (attributeData.AttributeType.FullName != JsonConverterAttributeFullName)
+                {
+                    return null;
+                }
+
+                Type converterType = new TypeWrapper((ITypeSymbol)attributeData.ConstructorArguments[0].Value, _metadataLoadContext);
+
+                if (converterType == null || converterType.GetConstructor(Type.EmptyTypes) == null || converterType.IsNestedPrivate)
+                {
+                    return null;
+                }
+
+                return $"new {converterType.GetUniqueCompilableTypeName()}()";
+            }
+
+            private void PopulateNumberTypes(MetadataLoadContextInternal metadataLoadContext)
+            {
+                Debug.Assert(_numberTypes != null);
```

Comment:
Do we not have nullable enabled for this library? This field is initialized inline, so these asserts seem a bit pointless.

Reply:
Yeah we can remove them.